### PR TITLE
Add initial draft of joint decorrelation

### PR DESCRIPTION
**Pull request by essink**

### CODE REVIEW

File: elephant/joint_decorrelation/joint_decorrelation.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    def fit(self, baseline_data=None, bias_filtered_data=None,
+            baseline_covmat=None, bias_filtered_covmat=None, verbose=False):
+
+        # check data shape compatibility
+        if baseline_covmat is None and bias_filtered_covmat is None:
+            if baseline_data.shape[0] != bias_filtered_data.shape[0]:
+                raise ValueError('`baseline_data` and `bias_filtered_data`'
+                                 'should have same size in the first dimension.')
+        else:
+            if not (baseline_covmat.shape[0] == baseline_covmat.T.shape[1] and
```
Comment (dizcza) [inline elephant/joint_decorrelation/joint_decorrelation.py]: you should compare `baseline_covmat.shape[0] == baseline_covmat.shape[1]`
then you don't need to unravel the matricies
just do `np.allclose(a, a.T)`

File: elephant/joint_decorrelation/joint_decorrelation.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                    np.all(np.ravel(bias_filtered_covmat) == np.ravel(bias_filtered_covmat.T))):
+                raise ValueError('`bias_filtered_covmat` should be a symmetric square'
+                                 'matrix.')
+            # check data shape compatibility
+            if baseline_covmat.shape != bias_filtered_covmat.shape:
+                raise ValueError('`baseline_covmat` and `bias_filtered_covmat` should have same'
+                                 'size.')
+
+        if baseline_covmat is None and bias_filtered_covmat is None:
+            c0 = np.cov(baseline_data)
```
Comment (dizcza) [inline elephant/joint_decorrelation/joint_decorrelation.py]: naming conventions!
add meaning to c's
clumsiness_of_mine1 = ...
same for `_fit` function

File: elephant/joint_decorrelation/joint_decorrelation.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            c1 = np.cov(bias_filtered_data)
+        else:
+            c0 = baseline_covmat
+            c1 = bias_filtered_covmat
+
+        return self._fit(c0, c1, verbose)
+
+    def _fit(self, c0, c1, verbose):
+
+        todss, fromdss, pwr0, pwr1 = dss0(c0, c1, keep1=self.n_max_components,
```
Comment (dizcza) [inline elephant/joint_decorrelation/joint_decorrelation.py]: these only-god-knows variables probably come from a paper
please specify this in the class doc. maybe add formulas too

File: elephant/joint_decorrelation/joint_decorrelation.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+        if len(components_to_discard) == 0:
+            components_to_discard = np.where(self.power_ratio > power_ratio_threshold)[0]
+
+        mask = np.ones(self.n_components, dtype=bool)
+        mask[components_to_discard] = False
+        return self._project(data, mask)
+
+    def _project(self, data, mask):
+        if np.all(mask) == True:
```
Comment (dizcza) [inline elephant/joint_decorrelation/joint_decorrelation.py]: is True

File: elephant/joint_decorrelation/joint_decorrelation.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        if data.shape[0] != self.todss.shape[0]:
+            raise ValueError('The first dimension of `data` should be equal to the number of channels.')
+        if not isinstance(components_to_discard, list):
+            raise ValueError('`components_to_discard` should be a list.')
+        if len(components_to_discard) > self.n_components:
+            raise ValueError('`components_to_discard` should be empty or a subset of all components.')
+        if len(components_to_discard) != 0 and max(components_to_discard) >= self.n_components:
+            raise ValueError('The indices in `components_to_discard` cannot exceed {} (`n_components`)'.format(self.n_components))
+
+        if len(components_to_discard) != 0 and power_ratio_threshold != None:
```
Comment (dizcza) [inline elephant/joint_decorrelation/joint_decorrelation.py]: is not None

File: elephant/joint_decorrelation/joint_decorrelation.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        mask[components_to_discard] = False
+        return self._project(data, mask)
+
+    def _project(self, data, mask):
+        if np.all(mask) == True:
+            return data
+        e_diag = np.zeros(self.n_components)
+        e_diag[mask] = 1
+        e = np.diag(e_diag)
+        return np.dot(np.dot(np.dot(data.T, self.todss), e), self.fromdss).T
```
Comment (dizcza) [inline elephant/joint_decorrelation/joint_decorrelation.py]: please break down the computation into intermediate steps
```
meaningful_variable = np.dot(data.T, self.todss)
yet_another_meaningful_variable = np.dot(meaningful_variable, e)
easy_to_understand = np.dot(yet_another_meaningful_variable, self.fromdss).T
return easy_to_understand
```

File: elephant/joint_decorrelation/joint_decorrelation.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+    def project_in(self, data, components_to_keep=[],
```
Comment (dizcza) [inline elephant/joint_decorrelation/joint_decorrelation.py]: this function is very similar to `project_out`
don't duplicate the code
think how to reuse the code
merge the functionality, if needed

**Comment by dizcza** [comment]
The authors of the PR, Simon and Junji, have started working on the joint decorrelation at the 1st Elephant hackathon. The code is based on the third-party repository (a link is missing).
The authors are not planning to continue working on this PR in the near future (say, half a year). Therefore, it can be safely closed without deleting the corresponding branch in INM-6/elephant. The PR can be reopened in the future if the authors decide to breeze the algorithm again.

**Comment by Moritz-Alexander-Kern** [comment]
See comment above.

Can be reopened at any time.

**Pull request closed without merging**